### ISSUE TITLE
Design Idea : Work order

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#857]
**Is your feature request related to a problem? Please describe.**
- Create the concept of a work order with clear goal and time period and requirement.

**Describe the solution you desire to see**
- Incorporate work order in the upcoming **Project** definition

**Describe alternatives you've considered**
- Code it as a sub-task in the current architecture

**Describe the rationale**
- Some maintenance, safety inspection, repair tasks may be considered as work orders
- Player (as a commander) may control and decide how often some recurrent work orders are carried out (e.g. once a day, or once every other day) 
- A facility working group (key stakeholders and decision makers) may be assembled (if commander allows it) to identify and approve each category of work orders

**Screenshots**
![Pflugerville_Blog-768x436](https://user-images.githubusercontent.com/1168584/233188939-e0515275-66e8-460f-ad54-d11babc3c242.jpg)

![image2-1-768x435](https://user-images.githubusercontent.com/1168584/233189557-1c0faa72-3e33-46c7-90e1-a1c7c40a332a.png)

**Specifications  (please complete the following information):**
 - pre-3.6.0 build 7959

**Additional context**
- Add visual elements and showcase the statistics of incomplete/complete work order
- Show bar and pie charts and a map showing the location and distribution of these orders

**Comment by mokun** [comment]
@Urwumpe @bevans2000 

This work order concept is not the same as the [Task Statistic Interface](https://github.com/mars-sim/mars-sim/issues/675).

It may be more similar to the [Task Board](https://github.com/mars-sim/mars-sim/issues/141) idea. 

But the difference is that we don't want to reuse the old `Task` model to accomplish a `Work Order`. We want something new and flexible and trackable.

**Comment by bevans2000** [comment]
What would be the Objectives?

**Comment by Urwumpe** [comment]
I agree. In that context, A work order seems to be more like a Project or Mission, not a task.

**Comment by mokun** [comment]
> What would be the Objectives?

One big objective is to break down further the jack-of-all-trade old idea of `Task` to do everything.

Work orders are usually recurrent repairs/inspections/tests that can be *scheduled* ahead of time. 

They are very *predictable* in nature. 

Currently, tasks are generated at runtime and no one know what a person will do next. The Schedule tab in a Person's Window only show what happen in the PAST and is after the fact. 

It'd say **predictability** is almost prerequisite for survival on Mars for the first generation of settlers.

So we need the idea of work orders to be implemented to make sense of our Mars simulation.

Btw, it doesn't seem to difficult to use JavaFX to create a side panel and the tables with search box. We may also try html with css. See below :  

![all-open-work-orders-dashboard-1536x878](https://user-images.githubusercontent.com/1168584/233193872-71d18aed-4089-47f7-a0fc-1ae9d8d6e217.png)

**Comment by Urwumpe** [comment]
I than disagree. We need underlying tasks to react to unpredicted (or at least too complex to predict) events. 

To get something like this, we would need an extension to the Mission framework and get a smaller type of Mission, that can be more fleeting, in the extreme consisting just of a single task.

**Comment by bevans2000** [comment]
Is it a variant on the existing Settlement Task backlog? A set of Tasks that are ready to action but some are repeating and follow a schedule? I can see we could add scheduled and repeating Task to the backlog logic. We would use the Future Event logic in the Settlement to add new Tasks to the backlog according to a schedule.
Or do we need a backlog of Tasks for the individual Person? 
Can these WorkOrders be done by anyone or do they belong to an individual?
Nevertheless we must remember that a Task is the Unit of work in the current simulation. That is what Workers do so if we introduce another concept it will have accommodate what we have already.

**Comment by mokun** [comment]
> Is it a variant on the existing Settlement Task backlog? 

![image](https://user-images.githubusercontent.com/1168584/233206477-f06ea0cb-37e0-46f4-abc3-cee809deb883.png)

Oh Yea I forgot we're now having a Task Backlog already.

![image](https://user-images.githubusercontent.com/1168584/233206830-2fdae0f9-6bfb-4400-8ba9-b82bbb645f40.png)

I can see we have maintain, review Mission, tend crop/fish, toggle resources, etc.

> some are repeating and follow a schedule? 

Yes. It's nice to see some of them in the form of a weekly schedule and this schedule preferably supports drag and drop for player to interact with.

**Comment by mokun** [comment]
> Or do we need a backlog of Tasks for the individual Person?

Currently, there is no assigned tasks for an individual. We could add that to the Command Dashboard

> Can these WorkOrders be done by anyone or do they belong to an individual?

Once we have set up a number of categories for these work orders, we may create an assignment system that will bring together a team that comprises personnel from various disciplines for more complex work orders.

But for simple tasks, I suppose only one primary and a backup member would be needed.

**Comment by mokun** [comment]
Currently, the Command Dashboard's Logistic tab offers a rudimentary way of adding new tasks and log that task : 

![image](https://user-images.githubusercontent.com/1168584/233208914-8d93b622-387e-48cf-a02c-077d086af647.png)

Can you further refine this UI for scheduling tasks ?

**Comment by bevans2000** [comment]
Yes it can be refined but what Scheduled tasks would need to be added? Then we can work on the ui

**Comment by mokun** [comment]
> Yes it can be refined but what Scheduled tasks would need to be added?

I'd like to see at least : 
1. toggle resources
2. building/vehicle maintenance
3. cooking, desert preparation
4. tend crop/fish
5. analyze map
6. review mission plans

See this recent [issue](https://github.com/mars-sim/mars-sim/issues/858).

Future tasks : 
1. safety inspection/trainings
2. analyze weather
3. analyze rock samples

**Comment by bevans2000** [comment]
> > Yes it can be refined but what Scheduled tasks would need to be added?
> 
> I'd like to see at least :
> 
> 1. toggle resources
> 2. building/vehicle maintenance
> 3. cooking, desert preparation
> 4. tend crop/fish
> 5. analyze map
> 6. review mission plans
> 
All these apart from 3 are already done as SettlementTasks. I can look to add that.

**Comment by mokun** [comment]
It seems Cooking is still not working even after [this](https://github.com/mars-sim/mars-sim/commit/f120adfdc408c462532fec20487ecafef01b7c50).

May be the replenishment rate is low. Nothing shows up in the JTable in Cooking Tab since sol 1.

![image](https://user-images.githubusercontent.com/1168584/234391753-8302be06-0281-4283-8dc2-1a83c2c066d4.png)

**Comment by bevans2000** [comment]
I have looked at moving cooking to the Settlement backlog yet so it's nothing to do with this issue. Basically no one is doing any Cooking tasks. I don't know why.

**Comment by mokun** [comment]
![image](https://user-images.githubusercontent.com/1168584/234398879-d41edd5d-7f53-4b72-96c8-0f09afd2408b.png)

On day 14, we have higher R-rate for meal but low rate for dessert.

May be the EatDrink class was discouraging people from eating meal/dessert ?!

**Comment by mokun** [comment]
Everyone in Zvenda is like Artem, who has never eaten a single serving of meal and dessert

![image](https://user-images.githubusercontent.com/1168584/234399647-39b89f67-879b-47d9-9680-6fb7637d7d71.png)

**Comment by bevans2000** [comment]
I had a quick look last night and basically there is never any ingredients for Cooking available. I don't know if there is a problem with the FoodProoduction so ingredients aren't being created.
But I don't understand the logic in the Cooking class as it holds a cache of ingredients that is only refreshed randomly 1 out of 5 times. Seems way more complicated than it needs to be. I don't understand why it can't just access the Settlement ResourceHolder directly to evaluate meals.
Maybe if we converted it to a SettlementTask we could simplify the approach by evaluating the potential meals as a single activity controlled by the Settlement meal times? We only need to prove there is 1 cookable meal to create an item on the backlog. If the backlog item is chosen to be completed then we can do the expensive evaluation of which meals can be cooked. 
It's a bit of change but is probably worth it.

**Comment by mokun** [comment]
>  I don't understand why it can't just access the Settlement ResourceHolder directly to evaluate meals.

Sure. One big problem was that if there are 4 kitchen competing for resources, how should it be handled and resources be distributed efficiently ? 

May be we can create a central pipeline/queue in such a way that each settlement would look for food ingredients for an estimated serving of meals at most once per millisol.

Say if it estimates at this moment, only 5 meals will be served, then the central pipeline will grab just enough ingredients for all 5 meals. 

Then say 2 kitchens will have cooks to fetch these meals. Say kitchen A will get ingredients for 2 meals. Kitchen B will get ingredient to make 3 meals.

This way, things will be a lot more streamline and much less chaotic in one millisol.

**Comment by mokun** [comment]
> Maybe if we converted it to a SettlementTask we could simplify the approach by evaluating the potential meals as a single activity controlled by the Settlement meal times?

Ok. I wasn't thinkin of `SettlementTask` earlier. May be it can be tweaked to incorporate that pipeline concept, can it ?

**Comment by mokun** [comment]
So as for dessert, there are two kinds. 

The "prepared" dessert needs to be made into servings of dessert.

On the other hand, the unprepared dessert can be consumed as just like dry food.

**Comment by bevans2000** [comment]
I think the root cause problem maybe that no Crops are reaching Harvest and hence no ingredients are created. I'm investigating but something doesn't look right

**Comment by bevans2000** [comment]
I was thinking of something similar and reusing what we have already built.
1. When the Settlement backlog is refreshed
2. Check there is Kitchen space available
3. Check it's meal time
4. Check at least 1 Meal has ingredients - check Resources directly against Settlement Resource Holder
5. Create a SettlementTask with the demand number equate to the number of Kitchen slots available

As it's done via the Backlog approach it is only evaluated infrequently and reused by many Person checks
Once a Cooking Task starts then the ingredients are taken directly from the Settlement

**Comment by mokun** [comment]
btw, there's no longer any issue with meal preparation. 

So now we're going back to the idea of work order in this thread.

I suppose we can accomplish this [one](https://github.com/mars-sim/mars-sim/issues/915) first and implement a new task for commander/chief. 

In future, once we flush out more clearly how to schedule tasks ahead, we'll go back to this thread.

**Comment by mokun** [comment]
@Urwumpe 

I just discovered we have `PersonFutureTask` hidden under /core/structure. 

@bevans2000 

you created it in 2023-11-18 according to its timestamp but you haven't put it into use.

I can that this really fit the idea of a work order or a scheduled task for those under a division (e.g. Resource department). 

What kind of task do you have in mind that you can readily use it now ? 

```java

/*
 * Mars Simulation Project
 * PersonFutureTask.java
 * @date 2023-11-18
 * @author Barry Evans
 */
package com.mars_sim.core.structure;

import com.mars_sim.core.data.RatingScore;
import com.mars_sim.core.events.ScheduledEventHandler;
import com.mars_sim.core.logging.SimLogger;
import com.mars_sim.core.person.Person;
import com.mars_sim.core.person.ai.task.util.BasicTaskJob;
import com.mars_sim.core.person.ai.task.util.MetaTaskUtil;
import com.mars_sim.core.person.ai.task.util.TaskFactory;
import com.mars_sim.core.person.ai.task.util.TaskJob;
import com.mars_sim.core.time.MarsTime;

/**
 * This class represents a future event that will create a task for a person.
 * It acts as a bridge between the ScheduledEvent and the TaskFactory.
 */
public class PersonFutureTask implements ScheduledEventHandler{

    private static final long serialVersionUID = 1L;

	private static SimLogger logger = SimLogger.getLogger(PersonFutureTask.class.getName());

    private transient TaskFactory factory;
    private String mtID;
    private Person person;

    /**
     * Creates a future event that will create a task for a person.

     * @param p Person wanting the Task
     * @param mt Factory to create the task
     */
    public PersonFutureTask(Person p, TaskFactory mt) {
        this.person = p;
        this.factory = mt;
        this.mtID = mt.getUpperCaseName();
    }

    /**
     * Gets the task factory for this event. This supports retrieving the factory
     * after serialization.   
     * @return
     */
    private TaskFactory getFactory() {
        if (factory == null) {
            factory = (TaskFactory) MetaTaskUtil.getMetaTask(mtID);
        }

        return factory;
    }

    /**
     * Gets the name of the event
     * 
     * @return name of the event
     */
    @Override
    public String getEventDescription() {
        return person.getName() + " - Future task of " + getFactory().getName();
    }

    /**
     * When executed this will create a pending Task and add it to the person.
     * 
     * @param currentTime This time of the event fired but not needed
     * @return Returns zero as this is not rescheduled.
     */
    @Override
    public int execute(MarsTime currentTime) {

//        logger.info(person, "Executing future task to add " + getFactory().getName() + ".");
        TaskJob job = new BasicTaskJob(getFactory(), RatingScore.ZERO_RATING);
        person.getTaskManager().addPendingTask(job, false);
        
        // Do not reschedule this event
        return 0;
    }
}

```

**Comment by bevans2000** [comment]
If we are not using them delete it

**Comment by mokun** [comment]
Don't delete it. 

I see we can make use of it to implement the idea of a scheduled work task. 

Can you provide an example on how to use it ?

**Comment by bevans2000** [comment]
We have source control, don't carry unused code. It adds code bloat. We already have a lot of unused code

**Comment by mokun** [comment]
I'm looking at GroupActivity since it also implements ScheduledEventHandler.

The method GroupActivity.createPersonActivity() is being used 3 times for 
1. Birthday party in `Person`
2. Council meeting in `Role`
3. New incoming settlers in `Resupply`

**Comment by bevans2000** [comment]
But we use ScheduledEventHandler in many places. This Person takes isn't needed. But just look eave it if you want. We don't need to change GroupActivity
